### PR TITLE
SLVS-709 Member override should do more than calling the base member

### PR DESCRIPTION
**Pull request by tvsonar**

### CODE REVIEW

File: src/SonarLint.CSharp/Rules/MemberOverrideCallsBaseMember.cs [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+            return hasReturnValue
+                ? blockSyntax == null
+                    ? expressionBodySyntax?.Expression
+                    : (GetSingleStatement(blockSyntax) as ReturnStatementSyntax)?.Expression
+                : (GetSingleStatement(blockSyntax) as ExpressionStatementSyntax)?.Expression;
+        }
+
+        private static StatementSyntax GetSingleStatement(BlockSyntax block)
+        {
+            return block.Statements.Count == 1 ? block.Statements.First() : null;
```
Comment (dbolkensteyn) [inline src/SonarLint.CSharp/Rules/MemberOverrideCallsBaseMember.cs]: @tvsonar this can throw a NPE while coding

File: src/SonarLint.CSharp/Rules/MemberOverrideCallsBaseMember.cs [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+            }
+
+            return true;
+        }
+
+        private static ExpressionSyntax GetExpressionToCheck(BlockSyntax blockSyntax, ArrowExpressionClauseSyntax expressionBodySyntax,
+            bool hasReturnValue)
+        {
+            return hasReturnValue
+                ? blockSyntax == null
```
Comment (dbolkensteyn) [inline src/SonarLint.CSharp/Rules/MemberOverrideCallsBaseMember.cs]: nested ternary is a bad idea

File: src/SonarLint.CSharp/Rules/MemberOverrideCallsBaseMember.cs [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+                propertySyntax.ExpressionBody == null)
+            {
+                // no getter
+                return true;
+            }
+
+            var memberAccess = GetExpressionToCheck(getAccessor?.Body, propertySyntax.ExpressionBody, true)
+                as MemberAccessExpressionSyntax;
+
+            if (memberAccess == null ||
```
Comment (dbolkensteyn) [inline src/SonarLint.CSharp/Rules/MemberOverrideCallsBaseMember.cs]: this check could be removed with no functional impact, the property check done in `IsBaseProperty` is more robust

File: src/SonarLint.CSharp/Rules/MemberOverrideCallsBaseMember.cs [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+
+            //check right:
+            var valueParameter = semanticModel.GetSymbolInfo(expressionToCheck.Right).Symbol as IParameterSymbol;
+            if (valueParameter == null || valueParameter.Name != "value" || !valueParameter.IsImplicitlyDeclared)
+            {
+                return false;
+            }
+
+            //check left:
+            var memberAccess = expressionToCheck.Left as MemberAccessExpressionSyntax;
```
Comment (dbolkensteyn) [inline src/SonarLint.CSharp/Rules/MemberOverrideCallsBaseMember.cs]: same thing here, this could be removed to simplify the code

File: src/Tests/SonarLint.UnitTest/TestCases/MemberOverrideCallsBaseMember.Fixed.cs [hunk 5]
```diff
@@ -0,0 +1,69 @@
+﻿using System;
```
Comment (dbolkensteyn) [inline src/Tests/SonarLint.UnitTest/TestCases/MemberOverrideCallsBaseMember.Fixed.cs]: @tvsonar I think it's preferable to have a dedicated input file to test the code fix.

Indeed, it's super annoying to have to change the Fixed.cs file when we only want to fix a false positive in the rule

The corner cases that we need for the rule detection and the fixer are also different: The fixer will always for example care about how comments and spaces are transformed, whereas we typically don't care at all about these for the detection.

**Comment by dbolkensteyn** [comment]
Also see https://jira.sonarsource.com/browse/SLVS-709

File: src/Tests/SonarLint.UnitTest/TestCases/MemberOverrideCallsBaseMember.Fixed.cs (hunk 5 above)
Comment (tvsonar) [inline src/Tests/SonarLint.UnitTest/TestCases/MemberOverrideCallsBaseMember.Fixed.cs]: That's true. However, I don't think we want to introduce one more test file for each analyzer, where we have a code fix. The ideal scenario is that we develop an analyzer, the test file contains all corner cases, we copy this file, remove the noncompliant code, and use the resulting file as the expected output of the codefix. And if either the analyzer or the codefix needs to be improved, then that will affect the other test file too.

File: src/SonarLint.CSharp/Rules/MemberOverrideCallsBaseMember.cs (hunk 3 above)
Comment (tvsonar) [inline src/SonarLint.CSharp/Rules/MemberOverrideCallsBaseMember.cs]: We can't remove it, consider the following:

    public class MyDerived : MyBase
    {
        private MyBase instance;
        public override int MyProperty1 // Noncompliant
        {
            get
            {
                return base.MyProperty1;
            }
        }

        public override int MyProperty2 // Compliant
        {
            get
            {
                return instance.MyProperty2;
            }
        }
    }

Comment (dbolkensteyn) [inline src/SonarLint.CSharp/Rules/MemberOverrideCallsBaseMember.cs]: please update the test case with this

File: src/SonarLint.CSharp/Rules/MemberOverrideCallsBaseMember.cs [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+            }
+
+            var methodSymbol = semanticModel.GetDeclaredSymbol(methodSyntax);
+
+            if (methodSymbol == null ||
+                !methodSymbol.IsOverride ||
+                methodSymbol.IsSealed ||
+                methodSymbol.OverriddenMethod == null ||
+                IgnoredMethodNames.Contains(methodSymbol.Name) ||
+                methodSymbol.Parameters.Any(p => p.HasExplicitDefaultValue))
```
Comment (dbolkensteyn) [inline src/SonarLint.CSharp/Rules/MemberOverrideCallsBaseMember.cs]: @tvsonar I would add this as well: `methodSymbol.OverriddenMethod.Parameters.Any(p => p.HasExplicitDefaultValue)`

it's not totally useless to want to make an optional parameter mandatory in a derived class, which is currently reported as redundant.

**Comment by dbolkensteyn** [comment]
LGTM, but see my comment about the use case where the base method has an optional parameter, which the derived class makes mandatory.
